### PR TITLE
Simple paginator

### PR DESCRIPTION
**Pull request by ADmad**
Added `SimplePaginator` which skips total count query. This is useful for paginating repositories with very large number of records where count query is very expensive or in case total count is unknown. 

Usage through `PaginatorComponent` would be like this:
```php
// in initialize()
$this->loadComponent('Paginator', [
    'paginator' => new \Cake\Datasource\SimplePaginator(),
]);
```

Refs https://github.com/cakephp/cakephp/issues/13418#issuecomment-526176622

### CODE REVIEW

File: src/Datasource/Paginator.php [hunk 1]
```diff
@@ -265,26 +263,31 @@ protected function buildParams(array $data)
...
             'page' => $page,
-            'current' => $data['numResults'],
+            'current' => $numResults,
             'count' => $count,
             'perPage' => $limit,
             'start' => $start,
             'end' => $end,
             'prevPage' => $page > 1,
-            'nextPage' => $count > ($page * $limit),
+            'nextPage' => $count === null ? true : ($count > ($page * $limit)),
```
Comment (markstory) [inline src/Datasource/Paginator.php]: Should determining the next/prev pages be moved into methods that implementations can override? Same for start/end?

File: src/Datasource/SimplePaginator.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.9.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Datasource;
+
+use Cake\Datasource\Exception\PageOutOfBoundsException;
+
+/**
+ * Simplified paginator which avoids query to get total count of records.
```
Comment (markstory) [inline src/Datasource/SimplePaginator.php]:
```suggestion
 * Simplified paginator which avoids potentially expensives queries 
 * to get the total count of records.
 * 
 * When using a simple paginator you will not be able to generate page numbers.
 * Instead use only the prev/next pagination controls, and handle 404 errors
 * when pagination goes past the available result set.
```

File: tests/TestCase/Datasource/SimplePaginatorTest.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ * @since         3.5.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Test\TestCase\Datasource;
+
+use Cake\Core\Configure;
+use Cake\Datasource\SimplePaginator;
+use Cake\ORM\Entity;
+
+class SimplePaginatorTest extends PaginatorTest
```
Comment (markstory) [inline tests/TestCase/Datasource/SimplePaginatorTest.php]: This seems efficient now, but will likely bite us as new tests are added and cause failures in a seemingly un-related test class. Another approach could be to move the shared test code into a trait that both tests include?

File: src/Datasource/Paginator.php (hunk 1 above)
Comment (ADmad) [reply src/Datasource/Paginator.php]: Okay, with do.

File: tests/TestCase/Datasource/SimplePaginatorTest.php (hunk 3 above)
Comment (ADmad) [reply tests/TestCase/Datasource/SimplePaginatorTest.php]: Sounds good, I make a trait.

File: src/Datasource/SimplePaginator.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.9.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Datasource;
+
+use Cake\Datasource\Exception\PageOutOfBoundsException;
```
Comment (josegonzalez) [inline src/Datasource/SimplePaginator.php]: This isn't actually used?

Comment (ADmad) [reply src/Datasource/SimplePaginator.php]: No, my phpcs precommit hook failed me.

**Comment by ADmad** [comment]
Updates done per per feedback.

**Comment by ADmad** [comment]
Anything else to be done here?

**Comment by ADmad** [comment]
The phpcs failures are unrelated.

**Comment by garas** [comment]
`$this->buildParams()` is called twice with same data

https://github.com/cakephp/cakephp/blob/c02272f84027b33784971d90cb41ac1d0abec506/src/Datasource/Paginator.php#L177-L179

**Comment by ADmad** [comment]
@garas  Fixed :slightly_smiling_face:

**Pull request merged**